gunrock/app/kcore/kcore_problem.cuh

```diff
@@ -0,0 +1,382 @@
+      GUARD_CU(initial_frontier.EnsureSize_(nodes, target));
+      GUARD_CU(empty_flag.EnsureSize_(1, target));
+      GUARD_CU(delete_bitmap.EnsureSize_(((nodes / 32) + 1), target));
+      GUARD_CU(to_be_deleted_bitmap.EnsureSize_(((nodes / 32) + 1), target));
+      // </TODO>
+
+      // Reset data
+      // <TODO> reset problem specific data, e.g.:
+
+      GUARD_CU(degrees.ForAll(
+            [sub_graph] __host__ __device__(int * degrees, const SizeT &pos) {
+              degrees[pos] = sub_graph.GetNeighborListLength(pos);
+            },
+            nodes, target, this->stream));
+
+      GUARD_CU(k_cores.ForEach([] __host__ __device__(SizeT & x) { x = (SizeT)0; },
+            nodes, target, this->stream));
+
+      GUARD_CU(empty_flag.ForAll(
+          [] __host__ __device__(SizeT * x, const VertexT &pos) { x[pos] = 0; },
+          1, target, this->stream));
+
+      GUARD_CU(to_be_deleted_bitmap.ForEach([] __host__ __device__(unsigned int & x) { x = 0u; },
+            ((nodes / 32) + 1), target, this->stream));
+
+      // </TODO>
+
+      return retval;
+    }
+  };  // DataSlice
+
+  // Set of data slices (one for each GPU)
+  util::Array1D<SizeT, DataSlice> *data_slices;
+
+  // ----------------------------------------------------------------
+  // Problem Methods
+
+  /**
+   * @brief k-core default constructor
+   */
+  Problem(util::Parameters &_parameters, ProblemFlag _flag = Problem_None)
+      : BaseProblem(_parameters, _flag), data_slices(NULL) {}
+
+  /**
+   * @brief k-core default destructor
+   */
+  virtual ~Problem() { Release(); }
+
+  /*
+   * @brief Releasing allocated memory space
+   * @param[in] target      The location to release memory from
+   * \return    cudaError_t Error message(s), if any
+   */
+  cudaError_t Release(util::Location target = util::LOCATION_ALL) {
+    cudaError_t retval = cudaSuccess;
+    if (data_slices == NULL) return retval;
+    for (int i = 0; i < this->num_gpus; i++)
+      GUARD_CU(data_slices[i].Release(target));
+
+    if ((target & util::HOST) != 0 &&
+        data_slices[0].GetPointer(util::DEVICE) == NULL) {
+      delete[] data_slices;
+      data_slices = NULL;
+    }
+    GUARD_CU(BaseProblem::Release(target));
+    return retval;
+  }
+
+  /**
+   * @brief Copy result distancess computed on GPUs back to host-side arrays.
+...
+   * \return     cudaError_t Error message(s), if any
+   */
+  cudaError_t Extract(
+      // <TODO> problem specific data to extract
+      SizeT *h_k_cores,
+      // </TODO>
+      util::Location target = util::DEVICE) {
+    cudaError_t retval = cudaSuccess;
+    SizeT nodes = this->org_graph->nodes;
+
+    if (this->num_gpus == 1) {
+      auto &data_slice = data_slices[0][0];
+
+      // Set device
+      if (target == util::DEVICE) {
+        GUARD_CU(util::SetDevice(this->gpu_idx[0]));
+
+        // <TODO> extract the results from single GPU, e.g.:
+        GUARD_CU(data_slice.k_cores.SetPointer(h_k_cores, nodes, util::HOST));
```

Comment:
h_k_cores is not necessarily a host-side pointer. It could be device side. See hits_problem.cuh::Extract()

Reply:
https://github.com/gunrock/gunrock/pull/842/commits/8dea4711205218851e6e9ef731c29ff6fdf74091